Building vignette on bringing your own data (#10)
fhdsl/Data_on_AnVIL: 02-bringing_own_data.Rmd

# (PART\*) Bringing Your Own Data {-}

```{r, include = FALSE}
ottrpal::set_knitr_image_path()
```

# Uploading from the Cloud {#uploading-cloud}

In this example, we'll upload some genomic data into AnVIL that is currently stored in the cloud (specifically, in a Google bucket).

::: {.dictionary}
**Buckets** are the name of the containers used to store files and objects on Google Cloud. Everything you store on Google Cloud _must_ be in a bucket. Each bucket has its own unique name and location (URI). When we move data files into AnVIL workspaces, we use the URI to tell AnVIL where the data should be stored. (We can also use a URI to tell AnVIL where to find the data we want to upload.)

You can read more about how data is saved in the cloud [here](https://support.terra.bio/hc/en-us/articles/360034335332-Understanding-data-in-the-Cloud)!
:::

> We're going to upload some fastq files for a SARS-CoV-2 sample. The bucket we're accessing contains 5 samples: two compressed fastq files, a fasta file for a SARS-CoV-2 reference genome, and two uncompressed fastq files. The bucket ID (URI) is `fc-80d0e1cd-61e9-472f-b1bd-c6a8223bd1cd`. For this activity, you will retrieve the two uncompressed fastq files and upload them into your workspace.

Reviewer: ```suggestion
We're going to upload some FASTQ files for a SARS-CoV-2 sample. The bucket we're accessing contains 5 samples: two compressed FASTQ files, a FASTA file for a SARS-CoV-2 reference genome, and two uncompressed FASTQ files. The bucket ID (URI) is `fc-80d0e1cd-61e9-472f-b1bd-c6a8223bd1cd`. For this activity, you will retrieve the two uncompressed FASTQ files and upload them into your workspace.
```

Once you have created a workspace, AnVIL will take you to the Workspace Dashboard. 

## Step Two: Provision a Jupyter environment

We will be using a terminal to transfer the files. You will need to provision a Jupyter environment in order to access a terminal window.

To access the menu for provisioning computing environments, double-click on the cloud-and-lightning-bolt icon in the right-hand menu.

```{r, echo=FALSE, fig.alt="Choosing the cloud-and-lightning-bolt icon will allow you to start the process of provisioning a computing environment."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3afa908e330_0_214#slide=id.g3afa908e330_0_214")
```

A new menu will pop up. Choose the "Settings" button under the Jupyter logo.

```{r, echo=FALSE, fig.alt="To provision a Jupyter environment, click on the settings button under the Jupyter logo."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3b11697e7a3_0_4#slide=id.g3b11697e7a3_0_4")
```

> The next menu will allow you to customize your Jupyter environment. For this exercise, the default settings are fine. If you know that your files are large and it will take some time to copy them into your Workspace bucket, make sure to deselect the "Enable Autopause" option. When you're ready, click the "Create" button.

Reviewer: Is it possible to provide an estimate for the minimum amount of  "large" that would need you to deselect the enable autopause option? Am I correct in assuming that you don't typically want to deselect that option because it could lead to larger costs? Is it worth noting that if so and maybe at the beginning of this section noting that besides storage costs, this method also has costs associated with the computing environment provisioning and use?

```

This will open a new tab in your browser window, but you are still in the same workspace! 

## Step Three: Run gcloud storage to copy/move files 

AnVIL Jupyter environments have the gcloud tools installed by default. This means you can perform command-line gcloud storage tasks to move/copy files between local storage and your workspace bucket (or any external bucket).

The command for this follows this pattern:

`gcloud storage cp <where_to_copy_data_from>/<file_name> <where_to_copy_data_to>`

You will need the names of both the bucket where the data is currently stored _and_ the bucket the data is being moved to.

The data is currently stored at `fc-80d0e1cd-61e9-472f-b1bd-c6a8223bd1cd`. Here's an image of the bucket contents as a reminder.

```{r, echo=FALSE, fig.alt="Image shows the contents of a Google bucket used in the SARS-CoV-2 on Galaxy activity."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3afa908e330_0_178#slide=id.g3afa908e330_0_178")
```

> AnVIL workspaces automatically comes with 50GB of storage (the persistent disk), as well as a unique workspace bucket.

Reviewer: ```suggestion
AnVIL workspaces automatically come with 50GB of storage (the persistent disk), as well as a unique workspace bucket.
```

You can click the clipboard icon on the right to copy the name of your Workspace Bucket.

```{r, echo=FALSE, fig.alt="The dashboard page of your new workspace shows information about the workspace on the right-hand side. You will want to copy the bucket name from here. This is the bucket where all information associated with this workspace is stored."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3afa908e330_0_205#slide=id.g3afa908e330_0_205")
```

**In order for `gcloud` to recognize the buckets, you need to add `gs://` to the front of their names.** So, for this example, the command to move the forward read file is

`gcloud storage cp gs://fc-80d0e1cd-61e9-472f-b1bd-c6a8223bd1cd/VA_sample_forward_reads.fastq gs://fc-970c8766-f42d-448e-9e4f-d86bbe13a065`

> Without the `gs://`, you will get an error message that the file URL does not exist!

Reviewer: URI instead of URL?

Author: Although actually, now that I think about it, the warning message specifically says URL

## Step Four: Verify that files have been transferred

> But how do you know if your files were successfully transferred?

Reviewer: ```suggestion
How do you know if your files were successfully transferred?
```

# (PART\*) Bringing Your Own Data {-}

```{r, include = FALSE}
ottrpal::set_knitr_image_path()
```

# Uploading From The Cloud {#uploading-cloud}

In this example, we'll upload some genomic data into AnVIL that is currently stored in the cloud (specifically, in a Google bucket).

> 

Reviewer 2: ```suggestion
<br><br>
```
Weird spacing 
<img width="174" height="85" alt="image" src="https://github.com/user-attachments/assets/4c1242a2-28b0-4da5-b3da-fde126596e7a" />

::: {.dictionary}
**Buckets** are the name of the containers used to store files and objects on Google Cloud. Everything you store on Google Cloud _must_ be in a bucket. Each bucket has its own unique name and location (URI). When we move data files into AnVIL workspaces, we use the URI to tell AnVIL where the data should be stored. (We can also use a URI to tell AnVIL where to find the data we want to upload.)

> You can read more about Google Cloud buckets [here](https://docs.cloud.google.com/storage/docs/buckets)!

Reviewer 2: Is there a Terra doc on this? I wonder if that could be more helpful/specific to our usecases?

Author: I didn't see anything when I looked previously, but I wouldn't be shocked if I used the wrong search terms. I'll look again.

:::

> We're going to upload some fastq files for a SARS-CoV-2 sample. The bucket we're accessing contains 5 samples: two compressed fastq files, a fasta file for a SARS-CoV-2 reference genome, and two uncompressed fastq files. The bucket ID (URI) is `fc-80d0e1cd-61e9-472f-b1bd-c6a8223bd1cd`. For this activity, you will retrieve the two uncompressed fastq files and upload them into your workspace.

Reviewer 2: ```suggestion
We're going to upload some fastq files for a SARS-CoV-2 sample. The bucket we're accessing contains 5 samples: 

* two compressed fastq files
* a fasta file for a SARS-CoV-2 reference genome
* two uncompressed fastq files. 

The bucket ID (URI) is `fc-80d0e1cd-61e9-472f-b1bd-c6a8223bd1cd`. 

For this activity, you will retrieve the two uncompressed fastq files and upload them into your workspace.
```

```{r, echo=FALSE, fig.alt="Image shows the contents of a Google bucket used in the SARS-CoV-2 on Galaxy activity."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3afa908e330_0_178#slide=id.g3afa908e330_0_178")
```

## Step One: Create your workspace

> The starting point for bringing your own data to AnVIL is the Workspace. Before you can do anything, you will need to create a workspace. Once you have logged into your AnVIL account, click on Workspaces in the left-side menu. You can open this menu by clicking the three line icon in the upper left hand corner.

Reviewer 2: Capitalization of workspace should be consistent. Might also link to the terra docs introducing workspaces

```

You should now see a pop-up window that lets you customize your new workspace. You will need to give your new workspace a unique name and assign it to a billing project. The "anvil-outreach" billing project is used here as an example, but you will not be able to assign it. You'll have to use one of your own billing projects. After filling out these two fields, click the "Quick Create Workspace" button to create a workspace without enabling sharing or additional security options.

```{r, echo=FALSE, fig.alt="You will need to give your new workspace a unique name and assign it to a billing project. The “anvil-outreach” billing project is used here as an example, but you won’t be able to use it. You’ll have to use one of your own. After filling out these two fields, click the “Quick Create Workspace” to create your workspace without enabling sharing or additional security options."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3afa908e330_0_197#slide=id.g3afa908e330_0_197")
```

> You can read about Authorization Domains for workspace security in [this article](https://support.terra.bio/hc/en-us/articles/8527464803739-How-to-set-up-and-use-an-Authorization-Domain) at the AnVIL Portal.

Reviewer 2: ```suggestion
You can read about Authorization Domains for workspace security in [this article](https://support.terra.bio/hc/en-us/articles/8527464803739-How-to-set-up-and-use-an-Authorization-Domain) in the Terra documentation.
```

```

AnVIL will now begin provisioning your environment. This will take a couple of minutes - when testing this vignette, it took between 2 and 3 minutes before the environment was ready. You'll know it's ready when the Jupyter icon on the menu has a green dot.

```{r, echo=FALSE, fig.alt="The Jupyter icon on the right-hand menu will have a green dot when your environment is ready to be used."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3b11697e7a3_0_12#slide=id.g3b11697e7a3_0_12")
```

Open a terminal window in your environment. You can access the terminal by clicking the icon below the Jupyter logo in the right-hand menu.

```{r, echo=FALSE, fig.alt="The terminal window can be opened by clicking on the icon underneath the Jupyter logo on the right-hand menu."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3b16c990b73_0_14#slide=id.g3b16c990b73_0_14")
```

This will open a new tab in your browser window, but you are still in the same workspace! 

> ## Step Three: Install gcloud storage

Reviewer 2: I believe gcloud tools are installed by default. Do you experience something different?

You will be asked where to create the directory. You can hit Enter to use the default.

Open a new bash shell in your terminal, or restart your shell using this command:

`exec -l $SHELL`

Finally, authenticate your gcloud by running the following:

`gcloud init`

::: {.notice}
Before downloading data using gcloud storage, verify your gcloud storage installation! Use the `ls` command to look at the buckets you can access.

Run `gcloud storage ls` to see all of the Cloud Storage buckets under the workspace project ID.

Run `gcloud storage ls -p [project name]` to list buckets for a specific project.

You can still move data into this Workspace even if the bucket isn't listed under this project ID!
:::

Once gcloud storage is installed, you can perform command-line gcloud storage tasks to move/copy files between local storage and your workspace bucket (or any external bucket).

## Step Four: Run gcloud storage to copy/move files 

Now that you've installed gcloud storage, you can copy your data files into the Workspace. The command for this follows this pattern:

`gcloud storage cp <where_to_copy_data_from>/<file_name> <where_to_copy_data_to>`

> You will need the names of both the bucket where the data is currently stored _and_ the bucket the data is being moved to.

Reviewer 2: Do we need some explanation of persistent disk here? Ie, why not just use `.` as the current location?

Author: Added!

Reviewer: Could possibly move the persistent disk descriptions/example into a details section or use a subheader for bucket and persistent disk sections. Felt a little scattered for me the way it's organized at the moment.

## Step Five: Verify that files have been transferred

But how do you know if your files were successfully transferred?

You can see any uploaded files by clicking the "Files" directory at the bottom left in the Data Tab.

```{r, echo=FALSE, fig.alt='Image shows a screenshot of the Workspace Data tab. The Files directory and link on the bottom left is highlighted.'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3b11697e7a3_0_24#slide=id.g3b11697e7a3_0_24")
```

If you click on the name of the file, you can also see the details of the file.

```{r, echo=FALSE, fig.alt='Image shows a screenshot of the details of the files we successfully transferred.'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3b11697e7a3_0_28#slide=id.g3b11697e7a3_0_28")
```

## Summary

- Create a workspace
- Provision a Jupyter environment
- Open terminal and install `gcloud` (if not done previously)
- run `gcloud storage cp <where_to_copy_data_from>/<file_name> <where_to_copy_data_to>`

> - verify that files have been transferred

Reviewer 2: ```suggestion
- Verify that files have been transferred
```
